testbed/tests/window/test_window.py

```diff
@@ -913,6 +915,200 @@ async def test_window_state_content_size_increase(
+        )
+
+        second_window.state = WindowState.MINIMIZED
+        await second_window_probe.wait_for_window(
+            "Setting to MINIMIZED state", state=WindowState.MINIMIZED
+        )
+        assert_window_event_triggered(second_window, second_window.on_hide)
+
+        second_window.state = state
+        await second_window_probe.wait_for_window(
+            f"Setting to final state of {state}", state=state
+        )
+        assert_window_event_triggered(second_window, second_window.on_show)
+
+    @pytest.mark.parametrize(
+        "visible_state1, visible_state2",
+        [
+            # We just need to test that setting a different visible state
+            # to a window already in visible-to-user(not minimized or not hidden)
+            # window state, doesn't trigger visibility events. Ideally, we would
+            # be testing for every possible pairings of the visible-to-user window
+            # states, but it would increase the runtime of the testbed.
+            (WindowState.NORMAL, WindowState.PRESENTATION),
+            (WindowState.MAXIMIZED, WindowState.FULLSCREEN),
+            (WindowState.FULLSCREEN, WindowState.MAXIMIZED),
+            (WindowState.PRESENTATION, WindowState.NORMAL),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "second_window_class, second_window_kwargs",
+        [
+            (
+                toga.Window,
+                dict(title="Secondary Window", position=(200, 150)),
+            )
+        ],
+    )
+    async def test_visibility_events_not_double_triggered(
+        second_window, second_window_probe, visible_state1, visible_state2
+    ):
+        """The window does not double trigger the on_hide() amd on_show() events."""
+        if not second_window_probe.supports_visibility_events:
+            pytest.xfail("This backend doesn't support visibility events.")
+        second_window.content = toga.Box(style=Pack(background_color=CORNFLOWERBLUE))
+        second_window.show()
+        second_window.on_show = Mock()
+        second_window.on_hide = Mock()
+
+        # visible-to-user = not minimized or not hidden
+        # not-visible-to-user state = minimized or hidden
+
+        second_window.state = WindowState.MINIMIZED
+        await second_window_probe.wait_for_window(
+            "Setting to MINIMIZED state", state=WindowState.MINIMIZED
+        )
+        assert_window_event_triggered(second_window, second_window.on_hide)
+
+        # The on_hide() event would not be triggered since the window is already in a
+        # not-visible-to-user(i.e., in minimized or hidden) state.
+        second_window.hide()
+        await second_window_probe.wait_for_window(f"Hiding {second_window.title}")
+        assert_window_event_triggered(second_window, expected_event=None)
+
+        second_window.show()
+        await second_window_probe.wait_for_window(f"Showing {second_window.title}")
+        if second_window_probe.show_unminimizes_window:  # For cocoa
```

Comment:
The behavior of `show()` on a minimized window should either be consistent on all platforms, or `show()`/`hide()` should raise an error if invoked on a minimized window. We shouldn't need a test workaround here - it's an area where we're in control of the API design.

If we're going to make the behavior consistent, I'd argue the Cocoa behavior is the better interpretation - if someone is calling "show()", they're indicating they want the window to be visible, not just minimized. But I suspect the "show/hide is an error if minimized" approach might be better as it's more explicit about intent.